pkg/gadgets/traceloop/tracer/tracer.go

```diff
@@ -0,0 +1,549 @@
+
+	r, ok := t.readers.Load(containerID)
+	if !ok {
+		return nil, fmt.Errorf("no perf reader for %q", containerID)
+	}
+
+	reader, ok := r.(*reader)
+	if !ok {
+		return nil, errors.New("the map should only contain *reader")
+	}
+
+	if reader.perfReader == nil {
+		log.Infof("reader for %v is nil, it was surely detached", containerID)
+
+		return nil, nil
+	}
+
+	err := readOverWritable(reader, func(record perf.Record, size uint32) error {
+		var cSyscallEvent C.struct_syscall_event_t
+		var cSyscallContEvent C.struct_syscall_event_cont_t
+
+		switch uintptr(size) {
+		case alignSize(unsafe.Sizeof(cSyscallEvent)):
+			cSyscallEvent = *(*C.struct_syscall_event_t)(unsafe.Pointer(&record.RawSample[0]))
+
+			event := &syscallEvent{
+				timestamp: uint64(cSyscallEvent.timestamp),
+				typ:       uint8(cSyscallEvent.typ),
+				contNr:    uint8(cSyscallEvent.cont_nr),
+				cpu:       uint16(cSyscallEvent.cpu),
+				id:        uint16(cSyscallEvent.id),
+				pid:       uint32(cSyscallEvent.pid),
+				comm:      C.GoString(&cSyscallEvent.comm[0]),
+				mountNsID: reader.mntnsID,
+			}
+
+			var typeMap *map[uint64][]*syscallEvent
+			switch event.typ {
+			case syscallEventTypeEnter:
+				event.args = make([]uint64, syscallArgs)
+				for i := uint8(0); i < syscallArgs; i++ {
+					event.args[i] = uint64(cSyscallEvent.args[i])
+				}
+
+				typeMap = &syscallEnterEventsMap
+			case syscallEventTypeExit:
+				// In the C structure, args is an array of uint64.
+				// But in this particular case, we used it to store a C long, i.e. the
+				// syscall return value, so it is safe to cast it to golang int.
+				event.retval = int(cSyscallEvent.args[0])
+
+				typeMap = &syscallExitEventsMap
+			default:
+				// Rather than returning an error, we skip this event.
+				// Indeed, I suspect this is caused because we copy the buffer while
+				// it is being written, so we will get uncomplete data, thus it is
+				// better to skip this event.
+				log.Debugf("type %d is not a valid type for syscallEvent, received data are: %v", event.typ, record.RawSample)
+				return nil
+			}
+
+			if _, ok := (*typeMap)[event.timestamp]; !ok {
+				(*typeMap)[event.timestamp] = make([]*syscallEvent, 0)
+			}
+
+			(*typeMap)[event.timestamp] = append((*typeMap)[event.timestamp], event)
+		case alignSize(unsafe.Sizeof(cSyscallContEvent)):
+			cSyscallContEvent = *(*C.struct_syscall_event_cont_t)(unsafe.Pointer(&record.RawSample[0]))
+
+			event := &syscallEventContinued{
+				timestamp: uint64(cSyscallContEvent.timestamp),
+				index:     uint8(cSyscallContEvent.index),
+			}
+
+			if cSyscallContEvent.failed != 0 {
+				event.param = "(Failed to dereference pointer)"
+			} else if uint64(cSyscallContEvent.length) == useNullByteLength {
+				// 0 byte at [C.PARAM_LENGTH - 1] is enforced in BPF code
+				event.param = C.GoString(&cSyscallContEvent.param[0])
+			} else {
+				event.param = C.GoStringN(&cSyscallContEvent.param[0], C.int(cSyscallContEvent.length))
```

Comment:
I think we need to remove non-printable characters. Otherwise, a pod can send terminal commands:
```
$ kubectl run -ti --rm --restart=Never --image=busybox shell -- sh
/ # echo -e '\033]0;Booo\007'  # Change the terminal window title
/ # echo -e "\a"               # Bell sound
/ # printf "\033c"             # Clear screen
```

Then, the terminal running the traceloop gadget will be affected.
```
$ kubectl gadget traceloop show 2d04aad11b6a26f0ac
```

You could use:
```
    event.param = strings.Map(func(r rune) rune {
        if unicode.IsPrint(r) {
            return r
        }
        return -1
    }, text)
```

Even without terminal commands, the output is difficult to read when I have new lines and non-printable characters mixed in the output.

Reply:
Instead of returning -1, I suggest to use a dot '.' as replacement for unprintable characters.